### PR TITLE
[Merged by Bors] - Tree hash caching and optimisations for Altair

### PR DESCRIPTION
**Pull request by michaelsproul**
## Proposed Changes

Remove the remaining Altair `FIXME`s from consensus land.

1. Implement tree hash caching for the participation lists. This required some light type manipulation, including removing the `TreeHash` bound from `CachedTreeHash` which was purely descriptive.
2. Plumb the proposer index through Altair attestation processing, to avoid calculating it for _every_ attestation (potentially 128ms on large networks). This duplicates some work from #2431, but with the aim of getting it in sooner, particularly for the Altair devnets.
3. Removes two FIXMEs related to `superstruct` and cloning, which are unlikely to be particularly detrimental and will be tracked here instead: https://github.com/sigp/superstruct/issues/5

### CODE REVIEW

**Review by realbigsean — APPROVED** [review]
I'm not too familiar with the tree hash caching functionality, but I looked through this pretty thoroughly and can't find any issues with it

File: consensus/types/src/participation_list.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    (len + BYTES_PER_CHUNK - 1) / BYTES_PER_CHUNK
+}
+
+pub fn leaf_iter(
+    values: &[ParticipationFlags],
+) -> impl Iterator<Item = [u8; BYTES_PER_CHUNK]> + ExactSizeIterator + '_ {
+    values.chunks(BYTES_PER_CHUNK).map(|xs| {
+        // Zero-pad chunks on the right.
+        let mut chunk = [0u8; BYTES_PER_CHUNK];
+        for (byte, x) in chunk.iter_mut().zip(xs) {
```
Comment (paulhauner) [inline consensus/types/src/participation_list.rs:50]: This is snazzy!

**Review by paulhauner — APPROVED** [review]
Also looks good to me!

File: consensus/types/src/participation_list.rs (hunk 1 above)
Comment (michaelsproul) [reply consensus/types/src/participation_list.rs:50]: Thanks! I was trying to write something that might get optimised into a `memcpy` by LLVM, or at least something with no bounds checks!

**Pull request closed without merging**